### ISSUE TITLE
[core] LanguageRegistry uses default class loader when invoking ServiceLocator

### ISSUE THREAD

**Issue by uhafner** [pmd/pmd#1377]
**Affects PMD Version:** 
all (6.8.0)

**Description:**

In Jenkins warnings plugin I am using [PMD core](https://github.com/jenkinsci/warnings-plugin/blob/master/src/main/java/io/jenkins/plugins/analysis/warnings/PmdMessages.java#L35) to read and print the descriptions of the warnings rules. I use the class `RuleSetFactory` to read the messages. In order to get the resources loaded I would like to provide Jenkins' classloader to the `RuleSetFactory`. This seems to be possible using the [4-arg constructor](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactory.java#L75). However, this currently does not work since the internal call to `LanguageRegistry` uses the [system default class loader](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageRegistry.java#L30). In order to get this running now, I need to use a "hack": I initialize the rules before Jenkins provides the correct class loader to my plug-in. This might change in the future.

**Comment by jsotuyod** [comment]
@uhafner I'm having a little trouble understanding your scenario.

You seem to be building a Jenkins plugin, which uses PMD. PMD is hence a direct dependency of your plugin and bundled with it (ie: present in the same classloader that loads and runs the plugin).

Jenkins is probably creating a separate classloader to run the plugin (as it's standard practice, to avoid conflicts on different versions of the same dependency), but that classloader probably references the Jenkins classloader as parent classloader to share the Jenkins API classes. Is this assumption correct?

In that scenario, anything in the Jenkins classloader is available to your plugin in the default classloader. You shouldn't need to specify a classloader yourself. What is exactly going on?

**Comment by uhafner** [comment]
If you don't specify a class loader when using Java's ServiceLoader then  'Thread.currentThread().getContextClassLoader()' will be used. This does not work in server applications like Jenkins. See [Jenkins Developer Documentation](https://jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/)

**Comment by jsotuyod** [comment]
@uhafner if I understand correctly, changing:

https://github.com/pmd/pmd/blob/14a375a51049132633bdb9b2258a77a25157d961/pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageRegistry.java#L30

to:

```java
ServiceLoader<Language> languageLoader = ServiceLoader.load(Language.class, getClass().getClassLoader());
```

would be enough, as Jenkins loads the classes from the proper classloader, it's simply not the default context one for the current thread, right?

**Comment by uhafner** [comment]
I'm not a class loader expert but from my understanding it should work. Kohsuke suggested to expose the class loader as a parameter in a library but this might not be needed.

**Comment by jsotuyod** [comment]
@uhafner ok. I don't think we can easily test that, but if we introduce the change, would you be willing to test against a snapshot build?

**Comment by uhafner** [comment]
Yes, of course. I already have a test case (integration test) that verifies this.

**Comment by adangel** [comment]
@uhafner The fix has been integrated now. Could you test, whether the problem is gone now?

Simply use PMD version 6.9.0-SNAPSHOT from

```
        <repository>
            <id>sonatype-nexus-snapshots</id>
            <name>Sonatype Nexus Snapshots</name>
            <url>https://oss.sonatype.org/content/repositories/snapshots</url>
            <releases>
                <enabled>false</enabled>
            </releases>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
```

Thanks!

**Comment by uhafner** [comment]
I’m in holidays until end of October. As soon as I am online again I will test the fix.

**Comment by uhafner** [comment]
I've tested my plug-in with the SNAPSHOT now, it works as expected. Thanks for fixing!

**Comment by jsotuyod** [comment]
@uhafner awesome! this was shipped in 6.9.0 on Sunday, you can use the release version now.

Thanks for testing this!

**Comment by uhafner** [comment]
Ah, even better, then I can integrate it right now 👍
